scripts/generate_drivers.py

```diff
@@ -0,0 +1,600 @@
+
+        read_func = f"""
+            uint{size}_t {device}_{name}_read(uint32_t *{device}_base)
+            {{
+                volatile uint32_t *control_base = {device}_base;
+                return METAL_{cap_device}_REGW(METAL_{cap_device}_{cap_name});
+            }}
+            """
+
+        rv.append(textwrap.dedent(read_func))
+
+    return '\n'.join(rv)
+
+
+def generate_metal_function(device: str, reg_list: t.List[Register]) -> str:
+    """
+    Generates the exported register access functions for
+    a given device and register list.
+
+    :param device: the name of the device
+    :param reg_list: the list of registers for the device.
+    :return:  the c code for the exported register access functions
+    """
+
+    rv: t.List[str] = []
+
+    for a_reg in reg_list:
+        name = a_reg.name.lower()
+        size = a_reg.width
+
+        write_func = f"""
+            void metal_{device}_{name}_write(const struct metal_{device} *{device}, uint{size}_t data)
+            {{
+                if ({device} != NULL)
+                    {device}->vtable.v_{device}_{name}_write({device}->{device}_base, data);
+            }}
+            """
+        rv.append(textwrap.dedent(write_func))
+
+        read_func = f"""
+            uint{size}_t metal_{device}_{name}_read(const struct metal_{device} *{device})
+            {{
+                if ({device} != NULL)
+                    return {device}->vtable.v_{device}_{name}_read({device}->{device}_base);
+                return (uint{size}_t)-1;
+            }}
+            """
+
+        rv.append(textwrap.dedent(read_func))
+
+    return '\n'.join(rv)
+
+
+def generate_metal_dev_drv(vendor, device, index, reglist):
+    """
+    Generate the driver source file contents for a given device
+    and register list
+
+    :param vendor: the vendor creating the device
+    :param device: the device
+    :param index: the index of the device used
+    :param reglist: the list of registers
+    :return: a string containing of the c code for the basic driver
+    """
+    template = string.Template(textwrap.dedent(METAL_DEV_DRV_TMPL))
+
+    return template.substitute(
+        vendor=vendor,
+        device=device,
+        cap_device=device.upper(),
+        index=str(index),
+        offsets=generate_offsets(device, reglist),
+        base_functions=generate_base_functions(device, reglist),
+        metal_functions=generate_metal_function(device, reglist),
+        def_vtable=generate_def_vtable(device, reglist)
+    )
+
+
+###
+# main
+###
+
+
+def handle_args():
+    """
+    :return:
+    """
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        "-o",
+        "--object-model",
+        type=argparse.FileType('r'),
+        help="The path to the object model file",
+    )
+
+    parser.add_argument(
+        "-d",
+        "--duh-document",
+        type=argparse.FileType('r'),
+        help="The path to the DUH document",
+        required=True
```

Comment:
why is this required? I though you could you this to just generate headers using just the object model?

Reply:
same with `--metal-dir`/`--bsp-dir`. It looks to me like these are just output directory args for the basic drivers and base headers respectively. why are they both required when you should be able to run either flow individually?

Reply:
because you might run them at the same time.

Reply:
but we want to be able to run them separately though. for the PIO block we only have access to the Object Model and we only want to generate the base header, how do we do that if we are also required to provide the DUH document?